### PR TITLE
Fix the return type of Trilinos's VectorReference::operator=.

### LINKED ISSUE #3726: Trilinos VectorReference constness
I reran cppcheck yesterday (patches are on their way) and found the following strange bit of code:

```cpp
/**
* Same as above but for non-const reference objects.
*/
const VectorReference &
operator = (const VectorReference &r);

/**
* Set the referenced element of the vector to <tt>s</tt>.
*/
const VectorReference &
operator = (const TrilinosScalar &s) const;
```

Should the first function return a `const VectorReference &`? The equivalent function in `PETScWrappers::VectorBase` does *not* return a `const` reference.

### PR DESCRIPTION
**Pull request by bangerth**
This was non-intuitive, though it did not make a difference given the particular
semantics of vector references. Fix it anyway.

Fixes #3726.

### CODE REVIEW

**Comment by bangerth** [comment]
The issue was so obscure that I didn't think a changelog entry was necessary. There should be no difference in observable behavior.

File: tests/trilinos/vector_reference_01.cc [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+  // copy VectorReference objects. note that operator= for these
+  // objects does not copy the*content* of the reference object, but
+  // assigns the value of the vector element referenced on the right
+  // to the vector element referenced on the left
+  (c = a) = b;
+  b = a;
+
+  deallog << static_cast<TrilinosScalar>(a) << std::endl; // should point to v(0)
+  deallog << static_cast<TrilinosScalar>(b) << std::endl; // should point to v(0)
+  deallog << static_cast<TrilinosScalar>(c) << std::endl; // should point to v(0)
```
Comment (drwells) [inline tests/trilinos/vector_reference_01.cc]: This comment does not match the output: we print `1.000`, not `0`, for `c`.

I think the the comment is wrong and the output is right. Lines 46 and 47 should do the following:
1. assign `c` to point to where `a` points (`c` points to entry 0)
2. assign `c` to point to where `b` points (`c` points to entry 1)
3. assign `b` to point where `a` points (`b` points to entry 0)

Is this correct?

File: tests/trilinos/vector_reference_01.cc [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+
+
+
+int main (int argc,char **argv)
+{
+  std::ofstream logfile("output");
+  deallog.attach(logfile);
+  deallog.threshold_double(1.e-10);
+
+  Utilities::MPI::MPI_InitFinalize mpi_initialization (argc, argv, testing_max_num_threads());
```
Comment (tjhei) [inline tests/trilinos/vector_reference_01.cc]: replace the four lines with ``MPILogInitAll log;``

File: tests/trilinos/vector_reference_01.cc [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+  v(0) = 0;
+  v(1) = 1;
+  v(2) = 2;
+
+  TrilinosWrappers::internal::VectorReference a (v(0));
+  TrilinosWrappers::internal::VectorReference b (v(1));
+  TrilinosWrappers::internal::VectorReference c (v(2));
+
+  // copy VectorReference objects. note that operator= for these
+  // objects does not copy the*content* of the reference object, but
```
Comment (tjhei) [inline tests/trilinos/vector_reference_01.cc]: missing space before "content"

File: tests/trilinos/vector_reference_01.cc [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+  deallog.attach(logfile);
+  deallog.threshold_double(1.e-10);
+
+  Utilities::MPI::MPI_InitFinalize mpi_initialization (argc, argv, testing_max_num_threads());
+
+
+  try
+    {
+      {
+        test ();
```
Comment (tjhei) [inline tests/trilinos/vector_reference_01.cc]: you don't need this try/catch stuff either, so please rip it out.

File: tests/trilinos/vector_reference_01.cc (hunk 1 above)
Comment (tjhei) [inline tests/trilinos/vector_reference_01.cc]: agreed.

**Comment by jppelteret** [comment]
Dammit. Sorry for the evidently premature merge :-/

**Comment by drwells** [comment]
@jppelteret these things happen; this is why we do reviews out in the open :)

**Comment by bangerth** [comment]
Fixed testcase in #3832. Sorry for the incomplete work :-(

**Comment by bangerth** [comment]
Oh, I see -- there is already #3830. I will let others figure out what to do next.

**Comment by jppelteret** [comment]
@bangerth No, its my fault for not being thorough enough when I looked at your first iteration.

**Comment by bangerth** [comment]
On 01/22/2017 09:07 AM, Jean-Paul Pelteret wrote:
> @bangerth <https://github.com/bangerth> No, its my fault for not being
> thorough enough when I looked at your first iteration.

It's never the reviewer's fault.

**Pull request merged**